src/vdom/index.js

```diff
@@ -10,19 +12,38 @@ export function isSameNodeType(node, vnode, hydrating) {
 	if (typeof vnode==='string' || typeof vnode==='number') {
 		return node.splitText!==undefined;
 	}
-	if (typeof vnode.nodeName==='string') {
-		return !node._componentConstructor && isNamedNode(node, vnode.nodeName);
+	const { _componentConstructor } = node;
+	const { nodeName } = vnode;
+	if (typeof nodeName==='string') {
+		return !_componentConstructor && isSameNodeName(node, nodeName);
 	}
-	return hydrating || node._componentConstructor===vnode.nodeName;
+	return hydrating || _componentConstructor===nodeName || isSameNodeConstructor(node, nodeName);
 }
 
 
 /** Check if an Element has a given normalized name.
-*	@param {Element} node
-*	@param {String} nodeName
+ *	@param {Element} node
+ *	@param {mixed} nodeName
+ */
+export function isSameNodeName(node, nodeName) {
```

Comment:
After adding `isSameNodeConstructor` (to be consistent with `isSameNodeType`, I felt that this should also be made consistent. I also couldn't understand the previous name as easily.